parser/parser.go

```diff
@@ -3067,7 +3068,58 @@ func (p *parser) parseGenDecl(keyword token.Token, f parseSpecFunction) *ast.Gen
 	}
 }
 
+func (p *parser) parseDecl(pkgName string, sync map[token.Token]bool) ast.Decl {
+	if p.trace {
+		defer un(trace(p, "Declaration"))
+	}
+	var f parseSpecFunction
+	pos := p.pos
+	switch p.tok {
+	case token.CONST, token.VAR:
+		f = p.parseValueSpec
+
+	case token.TYPE:
+		f = p.parseTypeSpec
+
+	case token.FUNC:
+		decl, call := p.parseFuncDeclOrCall(true)
+		if decl != nil {
+			if p.errors.Len() != 0 {
+				p.errorExpected(pos, "declaration", 2)
+				p.advance(sync)
+			}
+			return decl
+		} else {
+			decl = p.insertEntry(pos, &ast.ExprStmt{X: call})
+			if p.errors.Len() != 0 {
+				p.advance(sync)
+			}
+			return decl
+		}
+	default:
+		decl := p.insertEntry(pos)
+		if p.errors.Len() != 0 {
+			p.advance(sync)
+		}
+		return decl
+	}
+
+	return p.parseGenDecl(p.tok, f)
+}
+
+// ----------------------------------------------------------------------------
+// Source files
+
+func isOverloadOps(tok token.Token) bool {
+	return int(tok) < len(overloadOps) && overloadOps[tok] != 0
+}
+
 func (p *parser) parseFuncDecl() *ast.FuncDecl {
```

Comment:
I means that we keep the order of functions not changed so that we can see the smallest changes at code review stage.

The following code:

```go
func (p *parser) parseFuncDecl() *ast.FuncDecl {
    ...
}

func (p *parser) parseDecl(sync map[token.Token]bool) ast.Decl {
    ...
}
```

change into:

```go
func (p *parser) parseFuncDecl() *ast.FuncDecl {
    decl, _ := p.parseFuncDeclOrCall(false)
    return decl
}

func (p *parser) parseFuncDeclOrCall(allowCallExpr bool) (*ast.FuncDecl, *ast.CallExpr) {
    ...
}

func (p *parser) parseDecl(sync map[token.Token]bool) ast.Decl {
    ...
}

func (p *parser) insertEntry(pos token.Pos, stmts ...ast.Stmt) *ast.FuncDecl {
    ...
}
```

Reply:
insert parseDeclEx for support insert entry

Reply:
我的意思，每个函数的位置别改，这样的话 git diff 得到的修改也是最小化的。你现在把 parseDecl 挪到 parseFuncDecl 之前了，这样会显得修改了很多。